### PR TITLE
Old Code Revamp, Curved Rails, & Actor To Rail Fix

### PR DESCRIPTION
**Pull request by DonavinDraws**
### Features
* Curved Rails can be created and their curve control point can be moved in the viewport when a point is selected.
* Overlapping rail points is possible now
* ***Adding layers is now possible*** (limits to the amount of Play/DecoArea layers)
* **WIP** Deleting Layers
### Fixes
* Fixed crash with loading bone params
* Fixed how Actor To Rail Links are saved, as the previous method caused bugs with objects that moved along them automatically.
* Legacy ImGui Column code rewritten with the newer API's ImGui Table code.
* Tweaked the UI of Actor to Rail Link and Local Links window
* Fixed selecting Tile Units, Rails, and Global Links via the UI windows
* Fixed batch undo when deleting multiple objects

### CODE REVIEW

File: Fushigi/ui/widgets/CourseScene.cs [hunk 1]
```diff
@@ -2335,6 +2405,42 @@ private async Task<(CourseActor?, KeyboardModifier modifiers)> PickLinkDestInVie
...
+            CourseActorHolder actorArray, CourseAreaEditContext ctx)
+        {
+            var actors = actorArray.mActors.FindAll(x => x.mLayer == layer);
+            var confirm = await OperationWarningDialog.ShowDialog(mPopupModalHost,
+                "Deletion warning",
+                "Are you sure you want to delete " +
+                $"{layer}");
+
+            if (confirm == OperationWarningDialog.DialogResult.Cancel)
+                return;
```
Comment (jupahe64) [inline Fushigi/ui/widgets/CourseScene.cs:2419]: There's no need to display 2 OperationWarningDialogs, please **remove** this.
If there is no warnings simply display a Dialog without them (you could probably just move this code into the `else` block of `if (!noWarnings)` and ideally revert the if for easier readablity)

File: Fushigi/ui/widgets/CourseScene.cs [hunk 2]
```diff
@@ -2335,6 +2405,42 @@ private async Task<(CourseActor?, KeyboardModifier modifiers)> PickLinkDestInVie
             batchAction.Commit($"{IconUtil.ICON_TRASH} {actionName}");
         }
 
+        //TODO making this undoable
```
Comment (jupahe64) [inline Fushigi/ui/widgets/CourseScene.cs:2408]: How about making it **do** something in the first place?
I assume that's what you meant but I feel like the TODO comment should reflect that.

Also we should **not** expose functionality to users that isn't implemented yet.
At the VERY least put it in the title of the Warning Dialog => **Deletion warning (NOT IMPLEMENTED)**
or something like that

File: Fushigi/ui/widgets/CourseScene.cs [hunk 3]
```diff
@@ -2483,6 +2635,51 @@ private void SelectActorToAddLayer()
...
+
+                if (ImGui.BeginListBox("Select the layer you want to add the actor to.", ImGui.GetContentRegionAvail()))
+                {
+                    for (var i = 0; i < fileteredLayers.Count; i++)
+                    {
+                        var layer = i < 2 ? 
+                            fileteredLayers[i]+$" ({mLayersVisibility.Count(x => Regex.Replace(x.Key, @"\\d+", "").Contains(fileteredLayers[i]))}/10)":
+                            fileteredLayers[i];
+
+                        if(layer.Contains("(10/10)"))
```
Comment (jupahe64) [inline Fushigi/ui/widgets/CourseScene.cs:2665]: While this code may be very easy to read it's also very error prone
there should be a local variable `layerCount` and a (local) `const int MaxLayerCount = 10;`

Also `ImGui.BeginDisabled` has a condition parameter which I encourage using instead of putting both Begin and End in `if` clauses

File: Fushigi/ui/widgets/CourseScene.cs [hunk 4]
```diff
@@ -2381,33 +2487,79 @@ private async Task AddActorsWithSelectActorAndLayerWindow()
...
+
+            var layer = result.Value.layer;
+
+            if(layer == "PlayArea" || layer == "DecoArea")
+            {
+                var i = layer == "DecoArea" ? 0:1;
+                do
+                {
+                    i++;
+                } while (mLayersVisibility.Keys.Any(x => x == $"{layer}{i}"));
```
[the remark this answers is not in the document]
Comment (DonavinDraws) [reply Fushigi/ui/widgets/CourseScene.cs:2514]: Yeah, I didn't check because I assumed every level would start with one, also thanks, didn't know you could have a for loop without a condition.

File: Fushigi/ui/widgets/CourseScene.cs (hunk 2 above)
Comment (DonavinDraws) [reply Fushigi/ui/widgets/CourseScene.cs:2408]: I wanted to make deleting layers undoable, and now that you mention it, I do not know where that part went. I didn't want to implement the undo action half finished so I deleted the action commit, but I think I deleted the layer removal by mistake too. Regardless I figured out how to get the undo to work, I learned yesterday that assigning dictionaries to another variables did not work how I thought.

File: Fushigi/ui/widgets/CourseScene.cs [hunk 5]
```diff
@@ -2658,22 +2748,19 @@ private void SelectLayerToAdd()
                 {
                     for (var i = 0; i < fileteredLayers.Count; i++)
                     {
-                        var layer = i < 2 ? 
-                            fileteredLayers[i]+$" ({mLayersVisibility.Count(x => Regex.Replace(x.Key, @"\\d+", "").Contains(fileteredLayers[i]))}/10)":
-                            fileteredLayers[i];
+                        layerCount = mLayersVisibility.Count(x => RxLetters.Replace(x.Key, "") == fileteredLayers[i]);
+                        var layer = i < 2 ? $"{fileteredLayers[i]} ({layerCount}/{MaxLayerCount})" : fileteredLayers[i];
```
Comment (jupahe64) [inline Fushigi/ui/widgets/CourseScene.cs:2752]: Can't believe I didn't catch this before, this is REALLY bad code. It doesn't even work rn because you changed the order with the layerTypes but even if it was working, using the filter feature would still break it.
Also it's very hard to understand what the first 2 entries are supposed to be without the context of the previous commits.
Just compare the string directly and save everyone the headache.

File: Fushigi/ui/widgets/CourseScene.cs [hunk 6]
```diff
@@ -2658,22 +2748,19 @@ private void SelectLayerToAdd()
                 {
                     for (var i = 0; i < fileteredLayers.Count; i++)
                     {
-                        var layer = i < 2 ? 
-                            fileteredLayers[i]+$" ({mLayersVisibility.Count(x => Regex.Replace(x.Key, @"\\d+", "").Contains(fileteredLayers[i]))}/10)":
-                            fileteredLayers[i];
+                        layerCount = mLayersVisibility.Count(x => RxLetters.Replace(x.Key, "") == fileteredLayers[i]);
```
Comment (jupahe64) [inline Fushigi/ui/widgets/CourseScene.cs:2751]: Simply using StartsWith instead of removing all numbers with a Regex makes a lot more sense and should also be a lot faster

File: Fushigi/ui/widgets/LevelViewport.cs [hunk 7]
```diff
@@ -362,7 +362,7 @@ TileBfresRender CreateTileRendererForSkin(SkinDivision division, string skinName
             //Display skybox
             EnvironmentData.RenderSky(gl, this.Camera);
 
-            foreach (var actor in this.mArea.GetActors())
+            foreach (var actor in this.mArea.GetActors().OrderBy(x => x.mTranslation.Z))
```
Comment (jupahe64) [inline Fushigi/ui/widgets/LevelViewport.cs:365]: This code is completely unrelated to the commit, remove or explain
of course you can also put it in a seperate commit, I just feel like this **could** be a huge change in performance and UX so a commit message stating the intention would be very welcome
